File: php/examples/request_options.php
<?php

declare(strict_types=1);

require __DIR__ . '/../src/Auth.php';
require __DIR__ . '/../src/TokenSource.php';
require __DIR__ . '/../src/StaticTokenSource.php';
require __DIR__ . '/../src/ServiceClient.php';
require __DIR__ . '/../src/SdkConfig.php';
require __DIR__ . '/../src/DomainClients.php';
require __DIR__ . '/../src/SdkClient.php';

use TwoFinance\Sdk\RequestOptions;
use TwoFinance\Sdk\SdkClient;

$client = SdkClient::fromEnv();
$response = $client->analytics->post(
    '/analytics/candles:upsert',
    ['symbol' => 'BTC-USDT'],
    new RequestOptions(
        headers: ['X-Trace-ID' => 'trace-1'],
        idempotencyKey: 'candles-upsert-001',
        query: ['source' => 'sdk-example'],
        timeoutSeconds: 5.0,
        maxRetries: 1,
        page: 1,
        limit: 25,
    ),
);

echo 'response: ' . json_encode($response, JSON_THROW_ON_ERROR) . PHP_EOL;

File: cpp/examples/request_options.cpp
#include "twofinance/sdk_client.hpp"

#include <iostream>

int main() {
  twofinance::SdkConfig config;
  config.analytics_url = "https://analytics.example";

  twofinance::Transport transport = [](const twofinance::HttpRequest& request) {
    std::cout << request.method << " " << request.url << "\n";
    std::cout << request.headers.at("Idempotency-Key") << "\n";
    return twofinance::HttpResponse{200, "{\"ok\":true}"};
  };

  twofinance::RequestOptions options;
  options.headers["X-Trace-ID"] = "trace-1";
  options.idempotency_key = "candles-upsert-001";
  options.query["source"] = "sdk-example";
  options.page = 1;
  options.limit = 25;
  options.timeout_ms = 5000;
  options.max_retries = 1;

  twofinance::SdkClient client(config, transport);
  client.analytics.post("/analytics/candles:upsert", "{\"symbol\":\"BTC-USDT\"}", options);
  return 0;
}
